lib/LWP/Protocol/http.pm

```diff
@@ -89,7 +89,7 @@ sub _fixup_header
 
     # Extract 'Host' header
     my $hhost = $url->authority;
-    if ($hhost =~ s/^([^\@]*)\@//) {  # get rid of potential "user:pass@"
+    if ($hhost =~ s/^(.*)@//) {  # get rid of potential "user:pass@"
```

Comment:
This looks to me like the equivalent of:

Taking `$url->authority()` and then removing `$url->userinfo()`, but I'm not sure how that differs from just using `$url->host()`. That logic obviously predates this change. 🤔 

I tried to emit some debugging via the test suite, but it appears that there is no test coverage of when this condition is actually true.

Reply:
Yep, you might be right (and I'm 99% sure you're).
I might work on it as well, to simplify the logic of this, but this is pretty old code without UT coverage I don't want to broke something here. 
And to be honest, I'm not sure how the UT should look like to cover this case...
Anyway, at least I'm 100% sure that now it'll catch the credentials correctly, when they contain "@" symbol.

Reply:
For some reason request failed at this line if authorization contain "#" symbol:
https://github.com/libwww-perl/libwww-perl/blob/b8d96f01ea318365693a43ded28109a79ed941bd/lib/LWP/Protocol/http.pm#L214

URI lib cannot correctly parse next uri. For instance, next code:
```perl
use strict;
use warnings;
use v5.10;

use URI;

my $username = 'Str0ngUs3rW!th@#!$';
my $password = 'Str0ngP@$$w0rdW!th@#!$';
my $host = 'localhost';
my $port = '8080';

my $uri = URI->new("http://${username}:${password}\@${host}:${port}/path/to/file");
say $uri->scheme;
say $uri->host;
say $uri->port;
say $uri->path;
say $uri->authority;
say $uri->userinfo;
```
Return the next output:
```shell
$ perl test.pl
http

80

Str0ngUs3rW!th@
Str0ngUs3rW!th
```

I'd like to say it's a little broken...
Note: If I use uri_escape from URI::Escape before pass it into URI, everything worked as expected.
```shell
$ perl test.pl 
http
localhost
8080
/path/to/file
Str0ngUs3rW%21th%40%23%21%24:Str0ngP%40%24%24w0rdW%21th%40%23%21%24@localhost:8080
Str0ngUs3rW%21th%40%23%21%24:Str0ngP%40%24%24w0rdW%21th%40%23%21%24
```

I'll open a new issue in URI module....

Reply:
Nice find!